### PR TITLE
Support task mapping with setup teardown

### PR DESCRIPTION
**Pull request by dstandish**
hi @uranusjr here i have mapped setup / teardown _working_ ... if not in the prettiest way perhaps.  i wrote some dags to do some practical / visual testing.  i will work on making them "proper" tests this week. but if you have time to offer any guidance ...   or if you have other cases you think i should watch out for. .... thanks

### CODE REVIEW

**Comment by uranusjr** [comment]
The real meaningful change looks pretty straightfoward to me actually. Looking forward to this being cleaned up.

**Comment by dstandish** [comment]
> The real meaningful change looks pretty straightfoward to me actually. Looking forward to this being cleaned up.

Sorry didn't realize there was so much garbage in there from my `wip` commit @uranusjr ...  will clean it up... just pushed it quickly cus i didn't know if i'd have time to do a once over before turning in for my evening.

Anyway, meanwhile i'll just point out couple things
1. we didn't make the params `is_setup`, `is_teardown` and `on_failure_fail_dagrun` operator params, expecting that users would manage those attrs with the methods / decorator.  that is one way in which this is different from other attrs handled by  mapping / expand.  if it's a concern for you we could make them operator params just for consistency / compatibility
2. the method `as_teardown` is an instance method so that's why it ends up being helpful to be able to change trigger_rule on mappedoperator -- because it's not necessarily determined at time `expand` is called.

**Comment by dstandish** [comment]
ok ready for another look

File: airflow/decorators/base.py [hunk 1]
```diff
@@ -382,7 +382,15 @@ def expand(self, **map_kwargs: OperatorExpandArgument) -> XComArg:
...
         # to False to skip the checks on execution.
-        return self._expand(DictOfListsExpandInput(map_kwargs), strict=False)
+        if self.is_teardown:
+            if "trigger_rule" in self.kwargs:
+                raise ValueError("Trigger rule not configurable for teardown tasks.")
+            self.kwargs.update(trigger_rule=TriggerRule.ALL_DONE_SETUP_SUCCESS)
+        obj = self._expand(DictOfListsExpandInput(map_kwargs), strict=False)
+        obj.is_setup = self.is_setup
+        obj.is_teardown = self.is_teardown
+        obj.on_failure_fail_dagrun = self.on_failure_fail_dagrun
```
Comment (uranusjr) [inline airflow/decorators/base.py]: These very likely should be done in `_expand` instead to account for other call paths.

File: airflow/decorators/base.py [hunk 2]
```diff
@@ -382,7 +382,15 @@ def expand(self, **map_kwargs: OperatorExpandArgument) -> XComArg:
         prevent_duplicates(self.kwargs, map_kwargs, fail_reason="mapping already partial")
         # Since the input is already checked at parse time, we can set strict
         # to False to skip the checks on execution.
-        return self._expand(DictOfListsExpandInput(map_kwargs), strict=False)
+        if self.is_teardown:
+            if "trigger_rule" in self.kwargs:
+                raise ValueError("Trigger rule not configurable for teardown tasks.")
+            self.kwargs.update(trigger_rule=TriggerRule.ALL_DONE_SETUP_SUCCESS)
```
Comment (uranusjr) [inline airflow/decorators/base.py:388]: This should probably be in `__attrs_post_init__`? Also covers the non-mapped case.

File: airflow/migrations/versions/0128_2_7_0_add_is_setup_to_task_instance.py [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+airflow_version = "2.7.0"
+
+
+TABLE_NAME = "task_instance"
+
+
+def upgrade():
+    """Apply is_setup column"""
+    with op.batch_alter_table(TABLE_NAME) as batch_op:
+        batch_op.add_column(sa.Column("is_setup", sa.BOOLEAN(), nullable=False, server_default=sa.false()))
```
Comment (uranusjr) [inline airflow/migrations/versions/0128_2_7_0_add_is_setup_to_task_instance.py]:
```suggestion
        batch_op.add_column(sa.Column("is_setup", sa.Boolean(), nullable=False, server_default=sa.false()))
```

File: airflow/migrations/versions/0128_2_7_0_add_is_setup_to_task_instance.py [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+
+
+def downgrade():
+    """Unapply empty message"""
+    pass
```
Comment (uranusjr) [inline airflow/migrations/versions/0128_2_7_0_add_is_setup_to_task_instance.py]: Should implement this.

File: airflow/models/mappedoperator.py [hunk 5]
```diff
@@ -408,8 +390,23 @@ def trigger_rule(self) -> TriggerRule:
...
-        raise ValueError("Cannot set trigger_rule for mapped operator.")
+        self.partial_kwargs["trigger_rule"] = value
+
+    @property
+    def is_setup(self):
+        return self.partial_kwargs.get("is_setup", False)
+
+    @is_setup.setter
+    def is_setup(self, value):
+        self.partial_kwargs["is_setup"] = value
```
Comment (uranusjr) [inline airflow/models/mappedoperator.py]: I feel these setters aren’t really needed… There should be a way to supply those when a MappedOperator is created.

File: airflow/models/mappedoperator.py [hunk 6]
```diff
@@ -646,6 +643,8 @@ def unmap(self, resolve: None | Mapping[str, Any] | tuple[Context, Session]) ->
             # is prepended, and '__N' appended to deduplicate). This is hacky,
             # but better than duplicating the whole mangling logic.
             op.task_id = self.task_id
+            op.is_setup = kwargs.pop("is_setup", False)
+            op.is_teardown = kwargs.pop("is_teardown", False)
```
Comment (uranusjr) [inline airflow/models/mappedoperator.py]: `kwargs` is already passed to the operator’s constructor above, the pop calls should happen before that.

File: airflow/migrations/versions/0128_2_7_0_add_is_setup_to_task_instance.py (hunk 4 above)
Comment (dstandish) [reply airflow/migrations/versions/0128_2_7_0_add_is_setup_to_task_instance.py]: yup, thanks

File: airflow/models/mappedoperator.py (hunk 6 above)
Comment (dstandish) [reply airflow/models/mappedoperator.py]: yes, thank you, this was a "cleanup" that shouldn't have been cleaned up

**Comment by uranusjr** [comment]
The `COALESE` approach used previously doesn’t seem to sit well with all database setups; I converted it to use `CASE` instead and hope I did it correctly.

**Comment by dstandish** [comment]
Thanks a bunch @uranusjr 

I'll look over the changes

**Review by jedcunningham — APPROVED** [review]

**Review by uranusjr — APPROVED** [review]

**Comment by benbuckman** [comment]
This PR appears to have unintentionally(?), or at least without being documented in the changelog, changed the way `dag.test()` handles exceptions.
(See this line for example – https://github.com/apache/airflow/pull/32820/files#diff-62c8e300ee91e0d59f81e0ea5d30834f04db71ae74f2e155a10b51056b00b59bR2677 – `dag.test()` now returns a `DagRun` not `None`.)

This might actually help with writing unit tests for exceptions in DAGs – https://github.com/apache/airflow/discussions/32831 – I'm exploring that now.

Can this be added retroactively to the changelog for clarity?

**Comment by benbuckman** [comment]
I filed https://github.com/apache/airflow/issues/34490 to address this more directly.

**Pull request merged**